hotwrap.go

```diff
@@ -16,70 +22,71 @@ func main() {
 		log.Fatal("Failed to start hotwrap, no command specified in arguments ")
 	}
 
-
-	if os.Getenv("HOTWRAP_VERBOSE") != "" {
-
-	}
 	cmd := os.Args[1]
 	var args []string
 	if len(os.Args) > 1 {
 		args = os.Args[2:]
 	}
 
-	fdk.Handle(&hotWrap{
-		cmd:  cmd,
-		args: args,
-		env:  os.Environ(),
-	})
+	os.Stderr.WriteString(fmt.Sprintf(
+		"%v %v", cmd, strings.Join(args, " ")))
+	fdk.Handle(withError(cmd, args))
 
 }
 
-type hotWrap struct {
-	verbose bool
-	cmd     string
-	args    []string
-	env     []string
-}
-
-func (hw *hotWrap) logf(fmt string, args ...interface{}) {
-	if hw.verbose {
-		log.Printf(fmt, args...)
+func withError(execName string, execArgs []string) fdk.HandlerFunc {
+	f := func(ctx context.Context, in io.Reader, out io.Writer) {
+		err := runExec(ctx, fmt.Sprintf(
+			"%v %v", execName, strings.Join(execArgs, " ")), in, out)
+		if err != nil {
+			fdk.WriteStatus(out, http.StatusInternalServerError)
+			io.WriteString(out, err.Error())
+			return
+		}
+		fdk.WriteStatus(out, http.StatusOK)
 	}
+	return f
 }
 
-func (hw *hotWrap) Serve(ctx context.Context, r io.Reader, w io.Writer) {
-
-
-	hw.logf("Running '%s %s'", hw.cmd, strings.Join(hw.args," "))
-
-	baseEnv := hw.env
-
-	cmd := exec.Command(hw.cmd, hw.args...)
-	cmd.Env = baseEnv
-	cmd.Stdout = w
-	cmd.Stdin = r
-
-	stderr, err := cmd.StderrPipe()
-
-	if err !=nil {
-		log.Fatalf("Failed to open stderr pipe %s",err)
-
+func runExec(ctx context.Context, execCMDwithArgs string, in io.Reader, out io.Writer) error {
+	fctx := fdk.GetContext(ctx)
+	defer timeTrack(time.Now(), fmt.Sprintf("run-exec-%v", fctx.CallID()))
+	cancel := make(chan os.Signal, 3)
+	signal.Notify(cancel, os.Interrupt)
+	defer signal.Stop(cancel)
+	result := make(chan error, 1)
+	quit := make(chan struct{})
+	cmd := exec.CommandContext(ctx, "/bin/bash", "-c", execCMDwithArgs)
```

Comment:
this should work with any shell, ie `sh -c`, and not require bash, that will reduce the requirements and simplify the dockerfiles as well. I expect this program could be useful in the context of adding it to a user's custom image in order to execute their function without having to base on the hot wrap image, reducing the requirements to none would be nice (technically, a shell, which isn't a given either I guess). if a user wants specific shell behavior, they can set their default shell in their image and get it, but the requirement on bash here seems odd.

Reply:
are we using a shell to exec for signal handling? why are we using the shell at all (fork?)? we can exec the program in this process (as it was in the old version of this diff), are we handling the signals correctly or are there risks of zombie children here? I see goroutines but no killing the child process if we go over timeout (I am not sure if we're interested in that behavior)

Reply:
We use shell to exec because user may pass a shell script or whatever not only a binary. Also shell exec is necessary when user wants to use special chars (like things in awk, etc.).

Reply:
Okay, maybe adding bash was not that necessary but the idea is there: shell exec is necessary when we’re talking about command pipelines, shell scripts, etc. With the previous version it didn’t worked,